rfcs/2020-08-21-3092-apache-metrics-source.md

```diff
@@ -0,0 +1,235 @@
+BytesPerReq: 7406.93
+DurationPerReq: .366667
+BusyWorkers: 1
+IdleWorkers: 74
+Processes: 3
+Stopping: 0
+BusyWorkers: 1
+IdleWorkers: 74
+ConnsTotal: 1
+ConnsAsyncWriting: 0
+ConnsAsyncKeepAlive: 0
+ConnsAsyncClosing: 0
+Scoreboard: ________________________________________________________W__________________.....................................................................................................................................................................................................................................................................................................................................
+```
+
+I'll use this to generate the following metrics:
+
+- `apache.uptime_seconds` (counter)
+- `apache.total_accesses` (counter; extended)
+- `apache.total_kilobytes` (counter; extended)
+- `apache.total_duration` (counter; extended)
+- `apache.cpu_user` (gauge; extended)
+- `apache.cpu_system` (gauge; extended)
+- `apache.cpu_children_user` (gauge; extended)
+- `apache.cpu_children_system` (gauge; extended)
+- `apache.cpu_load` (gauge; extended)
+- `apache.requests_per_second` (gauge; extended)
+- `apache.bytes_per_second` (gauge; extended)
+- `apache.bytes_per_request` (gauge; extended)
+- `apache.duration_per_request` (gauge; extended)
+- `apache.busy_workers` (gauge)
+- `apache.idle_workers` (gauge)
+- `apache.processes` (gauge)
+- `apache.stopping` (gauge)
+- `apache.conns_total` (gauge)
+- `apache.conns_async_writing` (gauge)
+- `apache.conns_async_keepalive` (gauge)
+- `apache.conns_async_closing` (gauge)
+- `apache.scoreboard_waiting` (gauge)
+- `apache.scoreboard_starting` (gauge)
+- `apache.scoreboard_reading` (gauge)
+- `apache.scoreboard_sending` (gauge)
+- `apache.scoreboard_keepalive` (gauge)
+- `apache.scoreboard_dnslookup` (gauge)
+- `apache.scoreboard_closing` (gauge)
+- `apache.scoreboard_logging` (gauge)
+- `apache.scoreboard_finishing` (gauge)
+- `apache.scoreboard_idle_cleanup` (gauge)
+- `apache.scoreboard_open` (gauge)
+
+Metrics labeled `extended` are only available if `ExtendedStatus` is enabled
+for Apache. This is the default in newer versions (>= 2.4; released 2012), but
+purportedly [increases CPU
+load](https://www.datadoghq.com/blog/collect-apache-performance-metrics/#a-note-about-extendedstatus)
+so some users may turn it off. If it is off, they simply won't have those
+metrics published.
+
+I figure we probably don't want metrics for:
+
+- Load (should be handled by a `cpu` or similar metrics source)
+
+## Doc-level Proposal
+
+Users will be instructed to setup
+[`mod_status`](https://httpd.apache.org/docs/2.4/mod/mod_status.html) and
+enable
+[`ExtendedStatus`](https://httpd.apache.org/docs/2.4/mod/core.html#extendedstatus).
+
+The following additional source configuration will be added:
+
+```toml
+[sources.my_source_id]
+  type = "apache" # required
+  endpoint = "http://localhost/server-status?auto" # required, default
+  scrape_interval_secs = 15 # optional, default, seconds
```

Comment:
I just pulled that default from the `prometheus` source, but I'm happy to adjust it. I don't have strong opinions on what a good default is. I could see `1s` resulting in higher granularity metrics than people typically want to store.

Reply:
Well, part of our strategy is to be real-time and high-res, so I want to lean that way.

Reply:
:+1: I'm happy to default it to `1s`.

Reply:
There is a performance hit on Apache from hitting the mod_status endpoint, especially if extended is turned on (c.f. https://httpd.apache.org/docs/2.4/misc/perf-tuning.html). So we might want to be a little careful here. I am also not sure at the frequency of mod_status updates. Long time since I looked at that code.

Reply:
I was not aware this had a performance impact. We should investigate this before landing on 1s. My light investigation attempts have failed :/.

Reply:
Yeah, having not run Apache in a while, I don't have a good intuition for this.

Datadog does 15s and Telegraf does 10s but these are global defaults rather than specifically setup for scraping apache. This does raise the good point that, for the metrics collection sources that operate an interval, we should probably aim to use the same default for all for all of them to be less surprising.

Reply:
> we should probably aim to use the same default for all for all of them to be less surprising.

Agree

Reply:
Indeed - not only not surprising but it's hard to correlate samples with different intervals. :)

Reply:
I say we stick with `15s` since we are already using as the default for the prometheus scraper.

Reply:
Opened https://github.com/timberio/vector/issues/3563 to discuss this separately. We'll likely want to address the global default, etc, with that change.